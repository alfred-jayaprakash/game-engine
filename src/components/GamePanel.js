import React, {useState} from 'react';
import {
  Container,
  Row,
  Col,
  Card,
  CardImg,
  CardBody,
  Button,
  Input,
  ListGroup,
  ListGroupItem,
} from 'reactstrap';

const GamePanel = props => {
  const [answer, setAnswer] = useState ('');
  const [answers, setAnswers] = useState ([]);

  const onSubmit = e => {
    if (answer === '') {
      return console.log ('Answer cannot be empty');
    }
    if (answer.length > 10) {
      return console.log ('Answer length greater than 20');
    }

    if (answers.indexOf (answer) !== -1) {
      return console.log ('Word already guessed');
    }

    answers.push (answer);
    props.onAnswer (answer);
    setAnswer ('');
  };

  return (
    <div className="centered.form__box">
      <Card>
        <CardImg top width="100%" src={props.imgsrc} alt="Game Image" />
        <CardBody>
          <Container>
            <Row>
              <Col>
                <Input
                  type="answer"
                  name="answer"
                  id="userAnswer"
                  placeholder="Type your word"
                  onKeyDown={e => {
                    if (e.key === 'Enter') {
                      e.preventDefault ();
                      onSubmit (e);
                    }
                  }}
                  onChange={e => setAnswer (e.target.value)}
                  value={answer}
                />
                <Button color="success" onClick={onSubmit}>Submit</Button>
              </Col>
            </Row>
            <Row>
              <Col>
                {answers.map (answer => <Button key={answer}>{answer}</Button>)}
              </Col>
            </Row>
          </Container>
        </CardBody>
      </Card>
    </div>
  );
};

export default GamePanel;
