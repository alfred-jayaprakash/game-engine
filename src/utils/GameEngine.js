import io from 'socket.io-client';
const ENDPOINT = process.env.REACT_APP_SERVER_URL || '/';

let socket = null;

var GameEngine = (function () {
  return {
    connect: function () {
      socket = io (ENDPOINT);
      socket.on ('disconnect', () => {
        //Nothing for now
      });
    },
    join: function (username, room, callback) {
      socket.emit ('join', {username, room}, (error, data) => {
        callback (error, data);
      });
    },
    registerForRoomNotifications: function (callback) {
      socket.on ('room_data', data => {
        callback (data);
      });
    },
    sendGameStatus: function (gameData, callback) {
      socket.emit ('game_status', gameData, data => {
        callback (data);
      });
    },
    registerForGameStatus: function (callback) {
      socket.on ('game_status', data => {
        callback (data);
      });
    },
    registerForGameStateUpdates: function (callback) {
      socket.on ('state', data => {
        callback (data);
      });
    },
    disconnect: function () {
      //socket = null;
    },
  };
}) ();

export default GameEngine;
