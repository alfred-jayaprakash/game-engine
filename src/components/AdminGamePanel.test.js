import React from 'react';
import {render, fireEvent, screen, act, waitFor} from '@testing-library/react';
import userEvent from '@testing-library/user-event';
import {shallow} from 'enzyme';
import GamePanel from './GamePanel';
import GameEngine from '../utils/GameEngine';

let onAnswer;
let container;
const imageUrl = './1.jpg';
describe ('Snapshot tests', () => {
  beforeAll (() => {
    //ARRANGE
    GameEngine.connect ();
  });

  it ('should render correctly without issues', () => {
    onAnswer = () => {};
    const component = shallow (
      <GamePanel currentImage={imageUrl} onAnswer={onAnswer} />
    );

    expect (component).toMatchSnapshot ();
  });
});

describe ('Functional tests', () => {
  beforeAll (() => {
    //ARRANGE
    GameEngine.connect ();
    onAnswer = jest.fn ();
  });

  beforeEach (() => {
    container = document.createElement ('div');
    document.body.appendChild (container);
  });

  afterEach (() => {
    document.body.removeChild (container);
    container = null;
    onAnswer.mockClear ();
  });

  it ('initial screen should load correctly', () => {
    render (
      <GamePanel currentImage={imageUrl} onAnswer={onAnswer} />,
      container
    );

    //Make sure Submit button to be present
    expect (screen.queryByText ('Submit')).toBeInTheDocument ();
  });
});
