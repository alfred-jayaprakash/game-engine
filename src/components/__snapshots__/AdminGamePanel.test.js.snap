// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Snapshot tests should render correctly without issues 1`] = `
<div>
  <Row
    tag="div"
    widths={
      Array [
        "xs",
        "sm",
        "md",
        "lg",
        "xl",
      ]
    }
  >
    <Col
      tag="div"
      widths={
        Array [
          "xs",
          "sm",
          "md",
          "lg",
          "xl",
        ]
      }
    >
      <h4
        className="text-light"
      >
        Join Game ID: 
        123456
      </h4>
    </Col>
  </Row>
  <Row
    tag="div"
    widths={
      Array [
        "xs",
        "sm",
        "md",
        "lg",
        "xl",
      ]
    }
  >
    <Col
      tag="div"
      widths={
        Array [
          "xs",
          "sm",
          "md",
          "lg",
          "xl",
        ]
      }
    >
      <Button
        className="btn-lg btn-block"
        color="success"
        id="startbtn"
        onClick={[Function]}
        tag="button"
      >
        Start Game
      </Button>
    </Col>
  </Row>
  <Row
    tag="div"
    widths={
      Array [
        "xs",
        "sm",
        "md",
        "lg",
        "xl",
      ]
    }
  >
    <Col
      tag="div"
      widths={
        Array [
          "xs",
          "sm",
          "md",
          "lg",
          "xl",
        ]
      }
    >
      <Toast
        fade={true}
        isOpen={true}
        tag="div"
        transition={
          Object {
            "appear": true,
            "baseClass": "fade",
            "baseClassActive": "show",
            "enter": true,
            "exit": true,
            "in": true,
            "mountOnEnter": false,
            "onEnter": [Function],
            "onEntered": [Function],
            "onEntering": [Function],
            "onExit": [Function],
            "onExited": [Function],
            "onExiting": [Function],
            "tag": "div",
            "timeout": 150,
            "unmountOnExit": true,
          }
        }
      >
        <ToastHeader
          charCode={215}
          closeAriaLabel="Close"
          icon={
            <Spinner
              size="sm"
              tag="div"
              type="border"
            >
              Loading...
            </Spinner>
          }
          tag="strong"
          tagClassName="mr-auto"
          wrapTag="div"
        >
          <div>
            Waiting for other players to join ...
          </div>
        </ToastHeader>
        <ToastBody
          tag="div"
        >
          <Table
            responsiveTag="div"
            tag="table"
          >
            <tbody />
          </Table>
        </ToastBody>
      </Toast>
    </Col>
  </Row>
</div>
`;
